### PR TITLE
[FEATURE] New zip format

### PR DESCRIPTION
**Pull request by pblottiere**
## Description

In the context of the QEP https://github.com/qgis/QGIS-Enhancement-Proposals/issues/27 in which the goal is to provide an auxiliary storage, a new format *.qgz* is proposed here.

In fact, it's for now just a *zip* of the usual *.qgs* project file. But in a second phase (others PR will follow), the auxiliary storage among other things may be embedded within the zip project.

To do that, a new dependency has been introduced : *libzip*.

This PR is of course a proposition so comments/ideas/criticisms are obviously welcome!

Some tests have been added to check the current behavior.

## Checklist

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by NathanW2** [comment]
Nifty.  This does open a lot of options for us.  Thanks for working on it.

File: src/core/qgsproject.h
Comment (nyalldawson) [inline src/core/qgsproject.h]: I'm a little confused how this relates to QgsProject::fileName(). Am I right in concluding that for a zipped project, QgsProject::fileName() will be the path to the temporary unzipped qgs file, while zipFileName() will be the original path?

File: src/app/qgisapp.cpp
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: camelCase please :)

File: src/app/qgisapp.cpp
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: qt optimisation tip: toLower() for case insensitive matching is inefficient (requires that the whole string be converted to lowercase, the new lowercase string needs memory allocation, etc). Use .compare( ... , Qt::CaseInsensitive) instead, as that only needs to process as many characters in the original QString objects as are required before a mismatch is found.

File: src/app/qgisapp.cpp
Comment (nyalldawson) [inline src/app/qgisapp.cpp]: camelcase

File: src/core/qgsarchive.h
Comment (nyalldawson) [inline src/core/qgsarchive.h]: I think this should be left out of the base QgsArchive class, and QgsArchive should be left only with generic zip/unzip handling functions (nothing project specific - so reusable for other contexts). QgsProjectArchive could be a subclass with the extra project specific methods added.

File: src/core/qgsproject.cpp
Comment (nyalldawson) [inline src/core/qgsproject.cpp]: I think this needs ` | QIODevice::Truncate ` also (at least that was required in Qt4, not sure if it's been fixed in Qt5)

File: src/core/qgsproject.h
Comment (pblottiere) [inline src/core/qgsproject.h]: Yes that's it. 

Just to be sure we're on the same page, an example in Python console:

```` python
>>> project = QgsProject()
>>> project.unzip("/home/user/project.qgz")
True
>>> project.fileName()
'/tmp/QGIS3-XXXX/project.qgs'
>>> project.zipFileName()
'/home/user/project.qgz'
````

File: src/app/qgisapp.cpp
Comment (pblottiere) [inline src/app/qgisapp.cpp]: It's fixed, sorry for that...

File: src/app/qgisapp.cpp
Comment (pblottiere) [inline src/app/qgisapp.cpp]: Thanks for the tip! I updated the code accordingly.

File: src/core/qgsproject.h
Comment (nyalldawson) [inline src/core/qgsproject.h]: Ok. I'm not sold on that approach though - I think it'll mean there's a lot of extra logic which needs to be scattered throughout the code to handle this check if zip file name, if so use it, if not, use filename. E.g. just of the top of my head the project properties window shows file name, and there's an expression variable for file name. There's probably other use of QgsProject.fileName() too. Any code which handles translation of paths relative to the project path will also need to be updated, because those relative paths will be to the location of the qgz file.

I'd suggest leaving QgsProject.fileName() and .fileInfo() as always returning paths for the original user-location of the project (regardless of whether it's a qgs or qgz file), and adding some other accessor which always returns the path to the original or extracted .qgs file.

Comment (wonder-sk) [inline src/core/qgsproject.h]: Agreed with Nyall... I would even argue that client code should not have access to the temporary unpacked project file name - in theory right after load it could be removed again...?

File: src/core/qgsarchive.h
Comment (pblottiere) [inline src/core/qgsarchive.h]: You're right, it's done

File: src/core/qgsproject.h
Comment (wonder-sk) [inline src/core/qgsproject.h]: I am wondering if we need all these new methods in the public API - would it be possible to completely hide the details of zipping/unzipping inside QgsProject implementation? I would expect just one extra boolean "zipped" flag (with getter and setter): reading it would tell me if the project was zipped, setting it would determine whether the project should be zipped when written.

```python
project = QgsProject()
project.read('/home/xyz/project.qgz')
project.isZipped()  # returns True
```

**Comment by wonder-sk** [comment]
Very exciting stuff! Looking forward to the auxiliary storage!

File: src/core/qgsproject.h
Comment (pblottiere) [inline src/core/qgsproject.h]: @nyalldawson 

> I'd suggest leaving QgsProject.fileName() and .fileInfo() as always returning paths for the original user-location of the project (regardless of whether it's a qgs or qgz file), and adding some other accessor which always returns the path to the original or extracted .qgs file.

I agree with you, I'm gonna do that.

@wonder-sk 

> in theory right after load it could be removed again...?

It is quite true... I'm going to take a look at it.

File: src/core/qgsproject.h
Comment (pblottiere) [inline src/core/qgsproject.h]: Actually, it was my first approach. 

If @nyalldawson is agree with that too, I can go on this way?

**Comment by pblottiere** [comment]
@nyalldawson @wonder-sk Thank you for this first review, it's very helping!

I let you know!

File: src/core/qgsproject.h
Comment (m-kuhn) [inline src/core/qgsproject.h]: I'm also in favor of hiding this logic as much as possible from the API.

**Comment by m-kuhn** [comment]
This looks really promising!

**Comment by m-kuhn** [comment]
You can possibly add

    libzip-dev

to .travis.yml

File: .travis.yml
Comment (m-kuhn) [inline .travis.yml]: This shouldn't be required for the code layout build I think

Comment (pblottiere) [inline .travis.yml]: Indeed :)

**Comment by pblottiere** [comment]
I took into account your comments. Now we have something like this as @wonder-sk suggested:

``` python
project = QgsProject()
project.read('/home/xyz/project.qgz')
project.isZipped()  # returns True
```

Let me know what you think.

**Comment by pblottiere** [comment]
@nyalldawson @wonder-sk @m-kuhn @mhugo 

Do you have other comments about this PR? If not, would you considere to merge this PR?

Thank you!

File: src/core/qgsarchive.h
Comment (nyalldawson) [inline src/core/qgsarchive.h]: Should be virtual?

File: src/core/qgsarchive.h
Comment (nyalldawson) [inline src/core/qgsarchive.h]: Is this needed? Sipify should strip it out automatically.

File: src/core/qgsziputils.h
Comment (nyalldawson) [inline src/core/qgsziputils.h]: Yet -> already

File: src/core/qgsziputils.h
Comment (nyalldawson) [inline src/core/qgsziputils.h]: "...path to files to embed within..."

Bonus points if you add filles to the spell check file ;)

File: src/core/qgsarchive.h
Comment (3nids) [inline src/core/qgsarchive.h]: indeed

**Comment by pblottiere** [comment]
@nyalldawson I took into account your last comments. Thank's for the review!  Let me know.

File: src/core/qgsziputils.h
Comment (pblottiere) [inline src/core/qgsziputils.h]: > Bonus points if you add filles to the spell check file ;)

Done ;)

**Comment by mhugo** [comment]
@jef-n Hi. We have a new dependency here that does not exist in osgeo4w. What is the best way to do ? I've compiled and packaged it (visual studio 2015) [here](http://osgeo4w.oslandia.net/osgeo4w.test/x86_64/release/libzip/).
Let us know how we can help to have a functional windows night build once this PR merged

**Comment by pblottiere** [comment]
Is there other comments about the content of this PR? 

Because I would like to propose another PR soon with Auxiliary Storage but I need ZIP support for that.

Let me know what can I do to improve this PR to have something finally mergeable :)

**Comment by nyalldawson** [comment]
It's a +1 from me. But I'd like a second opinion for a change this significant.

**Review by nyalldawson — APPROVED** [review]

File: src/core/qgsarchive.h
Comment (m-kuhn) [inline src/core/qgsarchive.h]: Can this be private?

**Review by m-kuhn — APPROVED** [review]
+1, this is a very nice improvement

File: src/core/qgsarchive.cpp
Comment (m-kuhn) [inline src/core/qgsarchive.cpp]: What is this check here for?
That's probably happening if a) the file is empty, b) the file doesn't exit, c) the file is write protected d) some other process blocking the rm operation or similar.

In such cases, what's the "recovery scenario"?

Comment (pblottiere) [inline src/core/qgsarchive.cpp]: > What is this check here for?

I remove the file from the archive only if it's well removed from the filesystem but it's indeed maybe not a good idea... Because, as you said, I should have a "recovery scenario" to, for example, retry later in case of other process blocking the deletion... 

I should probably ignore that case considering that the aim is just to NOT zip the file the next time.

And by the way, I should check that the file is well contained within the archive before removing it...

**Comment by pblottiere** [comment]
Travis build is failing but it doesn't seem to be related to this PR.

**Comment by m-kuhn** [comment]
Probably an inconsistency between this branch and master. Rebase or run the script mentioned on travis manually to synchronize.

**Comment by pblottiere** [comment]
By the way, I rebased as @m-kuhn suggested and now everything seems green :)

**Comment by NathanW2** [comment]
NIce work. +1 to merge if others are happy.

**Review by rldhont — APPROVED** [review]
Great works and useful feature

**Comment by wonder-sk** [comment]
Looks good to me as well... +1 to merge the PR!

**Comment by nyalldawson** [comment]
***silent cheer**** (because everyone is too quiet in this room!)

**Comment by pblottiere** [comment]
Thank you for merging this PR :)

**Comment by nyalldawson** [comment]
@pblottiere thank you for doing all the hard work in this PR, so that @m-kuhn could steal all the credit by just hitting a single button...

**Comment by 3nids** [comment]
:floppy_disk: 🎉

**Comment by m-kuhn** [comment]
👏

**Comment by NathanW2** [comment]
Party time!

**Pull request merged**
